### PR TITLE
Dockerize FE and BE

### PR DESCRIPTION
**Pull request by OskarSchamardin**
Resolves issue #13. Add dockerfiles to FE and BE for running the whole project as if it were deployed on a server.

### CODE REVIEW

**Comment by Orzelius** [comment]
FYI the magic words are `closes #13`

File: README.md [hunk 1]
```diff
@@ -4,6 +4,58 @@ Loan and borrow!
...
+
+## Table of contents
+
+- [Important variables](#important-variables)
+
+- [Running the development services](#running-the-development-services)
+
+- [Stopping the development services](#stopping-the-development-services)
+
+- [Troubleshooting](#troubleshooting)
```
Comment (Orzelius) [inline README.md:21]: Please add these under `## Docker` or something similar, currently they look like unrelated paragraphs but I think they should be under a single heading.

File: be/README.md [hunk 2]
```diff
@@ -1,54 +1,66 @@
...
 
-## Running the app
+- [Troubleshooting](#troubleshooting)
 
-```bash
-# development
-$ npm run start
+## Important variables
+
+(no important variables specified)
```
Comment (Orzelius) [inline be/README.md:19]: There are some very important env variables in both fe and be They can be found in the `.env.example` files. Please link to them as well.

File: be/docker/be-dev.dockerfile [hunk 3]
```diff
@@ -0,0 +1,25 @@
+FROM node:lts-alpine
+
+WORKDIR /app
+
+# append node_modules to PATH
+ENV PATH /app/node_modules/.bin:$PATH
+
+# copy project files (docker-compose seems ignore '.dockerignore' files.
+# If you can make docker-compose ignore folders such as 'node_modules' please
+# make a pull request doing that (and replace these copy commands with 'COPY . ./').)
```
Comment (Orzelius) [inline be/docker/be-dev.dockerfile:10]: Will create an issue about this, let's deal with it in a separate pr

File: be/docker/be-dev.dockerfile [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+COPY nest-cli.json ./
+COPY package-lock.json ./
+COPY package.json ./
+COPY schema.gql ./
+COPY src ./src/
+COPY test ./test/
+COPY tsconfig.build.json ./
+COPY tsconfig.json ./
+
+RUN npm i
```
Comment (Orzelius) [inline be/docker/be-dev.dockerfile:23]: Please replace with `npm ci` [npm-ci docs](https://docs.npmjs.com/cli/v7/commands/npm-ci) in both of the projects

File: fe/docker/www-dev.dockerfile [hunk 5]
```diff
@@ -0,0 +1,17 @@
+FROM node:lts-alpine
+
+WORKDIR /app
+
+# append node_modules to PATH
+ENV PATH /app/node_modules/.bin:$PATH
+
+# install dependencies
+COPY package.json ./
+COPY package-lock.json ./
```
Comment (Orzelius) [inline fe/docker/www-dev.dockerfile:10]: Please also copy the gitignored `.env` file

File: be/docker/be-dev.dockerfile [hunk 6]
```diff
@@ -0,0 +1,25 @@
+FROM node:lts-alpine
+
+WORKDIR /app
+
+# append node_modules to PATH
+ENV PATH /app/node_modules/.bin:$PATH
```
Comment (Orzelius) [inline be/docker/be-dev.dockerfile:6]: You'll also need to add the variables from (gitignored) `.env` to the container,

**Review by Orzelius — CHANGES_REQUESTED** [review]
I'm really impressed with how fast you got this done, but I'd like some small things changed

File: README.md (hunk 1 above)
Comment (OskarSchamardin) [reply README.md]: Fixed in commit 2f5fa9bd5ef486222acdb37cbfcda00f53a76282.

File: be/docker/be-dev.dockerfile (hunk 3 above)
Comment (Orzelius) [reply be/docker/be-dev.dockerfile:10]: Or you know what, create the issue yourself and assign to yourself/Sten

File: be/README.md (hunk 2 above)
Comment (OskarSchamardin) [reply be/README.md]: Fixed in new commit: 54e1249614707588a8a1562d9ac515df42d9ec53.

File: be/docker/be-dev.dockerfile (hunk 3 above)
Comment (OskarSchamardin) [reply be/docker/be-dev.dockerfile:10]: Please link the issue here when it gets created.

File: fe/docker/www-dev.dockerfile (hunk 5 above)
Comment (OskarSchamardin) [reply fe/docker/www-dev.dockerfile]: Fixed in new commit: 8ffdf9d8def4e27d69b9dfaeaf6d21dcdc414892.

File: be/docker/be-dev.dockerfile (hunk 6 above)
Comment (OskarSchamardin) [reply be/docker/be-dev.dockerfile:6]: Fixed in new commit: 8ffdf9d8def4e27d69b9dfaeaf6d21dcdc414892.

File: be/docker/be-dev.dockerfile (hunk 3 above)
Comment (OskarSchamardin) [reply be/docker/be-dev.dockerfile:10]: Created issue #45.

File: be/docker/be-dev.dockerfile [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+# copy project files (docker-compose seems ignore '.dockerignore' files.
+# If you can make docker-compose ignore folders such as 'node_modules' please
+# make a pull request doing that (and replace these copy commands with 'COPY . ./').)
+COPY .env.example ./
+COPY .env ./
+COPY .eslintrc.js ./
+COPY .prettierrc ./
+COPY nest-cli.json ./
+COPY package-lock.json ./
```
Comment (Kasparsu) [inline be/docker/be-dev.dockerfile:16]: I mean at this point wouldn't it be easier to just copy all and then delete some?

Comment (Kasparsu) [reply be/docker/be-dev.dockerfile:16]: Also this just copys them at build so if you need to change anything you will need to rebuild docker image. Use docker-compose.yml volumes instead. just mount it to correct place and all good.

Comment (Kasparsu) [reply be/docker/be-dev.dockerfile:16]: you are kinda mixing up dev docker and production docker and ci docker here I am feeling like.

**Review by Kasparsu — CHANGES_REQUESTED** [review]

File: be/docker/be-dev.dockerfile (hunk 7 above)
Comment (Orzelius) [reply be/docker/be-dev.dockerfile:16]: Yep this was supposed to be a dev-docker setup task (I don't know who would use it though as all the things run quite well locally other than the db)

Comment (OskarSchamardin) [reply be/docker/be-dev.dockerfile:16]: > Also this just copys them at build so if you need to change anything you will need to rebuild docker image. Use docker-compose.yml volumes instead. just mount it to correct place and all good.

The thing about docker volumes is that they don't want to work with existing folders (for me) and newly created volumes have a different owner and permissions, which means that reading and writing to those folders becomes a pain (on Linux). Maybe i'm doing it wrong (if so please let me know).

Any new ideas are welcome.

**Comment by OskarSchamardin** [comment]
Just force pushed a change. This removes the accidental modifications I made to the back end's `package-lock.json` file by running `npm i` at the time. This removal prevents 13000 deletions to the file (according to git).

File: be/docker/be-dev.dockerfile (hunk 7 above)
Comment (OskarSchamardin) [reply be/docker/be-dev.dockerfile]: Fixed with commit 16510fb986410b1627482c137abd79ed92f2e5b0.

**Review by Kasparsu — APPROVED** [review]

**Review by Orzelius — APPROVED** [review]

**Pull request merged**